### ISSUE TITLE
Remove get-default

### ISSUE THREAD

**Issue by Ambrevar** [atlas-engineer/next#654]
Since 66d53a9f9a5b56f57668e65cf73c11e96f2eff22 we don't use get-default anymore.

The function is rather hackish since it's unclear whether the initform should be evaluated or not (see the big comment in its definition).

The `find-slot` function can go with it.

**Comment by Ambrevar** [comment]
There is one area where `get-default` is useful: to get the default value of a slot when subclassing buffer / window / browser in the user config.

For instance I have
```lisp
(define-configuration browser
  ((search-engines (append my-search-engines (get-default 'browser 'search-engines)))))
```

Suggestion 1: Replace get-default by

```lisp
(getf (mopu:slot-properties 'gtk-browser 'search-engines) :initform)
```

Suggestion 2: Extend define-configuration to substitute a special symbol, say
%slot-default, to the parent slot initform.  Example:

```lisp
(defmacro define-configuration (super &body slots)
  (let* ((name (intern (str:concat "USER-" (symbol-name super))))
         (configured-class (intern (str:concat "*" (symbol-name super) "-CLASS*")))
         (super (intern (str:concat (symbol-name *renderer-class*) "-" (symbol-name super)))))
    `(progn
       (defclass ,name (,super)
         ,(loop for slot in (car slots)
                for initform = (getf (mopu:slot-properties super (car slot)) :initform)
                collect (list (car slot) :initform `(funcall (lambda (next-user:%slot-default) ; TODO: Need to export %slot-default
                                                               ,(cadr slot))
                                                             ,initform))))
       (setf ,configured-class ',name))))
       
       
;; init.lisp
(define-configuration browser
  ((search-engines (append my-search-engines %slot-default))))
```

Thoughts?

**Comment by jmercouris** [comment]
That's a most interesting idea. What do you think about:

```
(define-configuration browser
  ((search-engines (append my-search-engines (browser search-engines))))
```

because `browser` is technically the super class. We could also have a form like (append-to-default xyz) that we handle.

I also like Suggestion 1 as well. More thoughts :-D?

**Comment by Ambrevar** [comment]
> ```
> (define-configuration browser
>   ((search-engines (append my-search-engines (browser search-engines))))
> ```
>
> because `browser` is technically the super class.

How would `(browser search-engines)` work?  What is `browser` and what
is `search-engine`?

Note that the superclass is not instantiated, so we can't access the
instantiated default value.


> We could also have a form like (append-to-default xyz) that we handle.

You mean we handle it in the macro?  Can you elaborate?

> I also like Suggestion 1 as well. More thoughts :-D?

Actually suggestion 1 could be used nonetheless by users who want to
define their classes manually.

**Comment by jmercouris** [comment]
`(browser search-engines)`, browser is expanded by the macro to `gtk-browser`, and search-engines is the accessor for search-engines, we can use moptilities to get the initform and append to it.

With regards to (append-to-default xyz), I did mean handling it in the macro. It would do the same as described above, it would just be a convenience.

As per Suggestion 1, it is indeed not mutually exclusive!

**Comment by Ambrevar** [comment]
The problem with expanding `(browser search-engines)` at macro time is that it's brittle.
Say the user flets `browser` to something, then their `(browser ...)` call will lead to something confusing :/

It's also hard to parse.  How do you parse `'(browser search-engines)` (notice the quote) or `(quote (browser search-engines))`?

The benefit of using a lambda as I suggested above is that it performs lexical substitution, so it behaves exactly like how any programmer would expect.  Plus implementation is trivial.

**Comment by Ambrevar** [comment]
Then we need to document define-config.

**Comment by jmercouris** [comment]
The lambda makes sense then, I also like the get-default append method, seems more straightforward, it does however require the user to know some more things. We can have both available, thoughts?

**Comment by jmercouris** [comment]
Also yes, define-configuration, and the manual in general could use a nice freshening up :-)

**Comment by Ambrevar** [comment]
Which get-default?
The point of this issue was to remove get-default! :)
(Because it's brittle and broken in so many ways.)

**Comment by Ambrevar** [comment]
To clarify, the mopu version does not fix the evaluation issue I mentioned in the first comment.  So if the initform contains something that's meant to be evaluated it will not do what the user thinks, which is wrong.

What I meant with "replace get-default with mopu" is for advanced users only who would redefine their own get-default function.  I don't think we should export such a broken function to the users.

**Comment by jmercouris** [comment]
I don’t see how the moptilities version is broken, it works, just the behavior may not be what the user expects :-)

To answer the above, yes let’s replace get default with moptilities

**Comment by jmercouris** [comment]
Actually, on second thought, I change my vote to removing get-default and extending the define-configuration macro as you have illustrated

**Comment by Ambrevar** [comment]
Thanks.
So I'll merge the snippet I've posted above and remove get-default.

**Comment by Ambrevar** [comment]
Done in  7b66b3fe539c3b61d22cb523cab70b3906af7257.

**Comment by jmercouris** [comment]
Thank you!
